### PR TITLE
if f1 is v1 and f2 is v2, then f3 must be [opt]

### PR DESCRIPTION
**Pull request by kaitlyn-sharo**
Does not require field3 if conditions are met by field1 and field2. Does not consider requirements of fields at all, only checks:
- if the values of f1 and f2 are provided and match v1 and v2 respectively, if the value of f3 is provided then it must be one of [options]

Is expandable to more than 2 conditions and more than 1 requirement. Check the README for formatting advice.

### CODE REVIEW

File: api_files/dependencies.json [hunk 1]
```diff
@@ -60,5 +60,20 @@
...
+    "deepLinks": {
+        "state.characters[].demographics": [
+            {
+                "condition": {
+                    "rank": "O-3",
+                    "military_branch": "US Navy"
+                },
+                "requirement": {
+                    "rank_title": [
+                        "Lieutenant"
```
Comment (brianpippin) [inline api_files/dependencies.json:73]: So I was testing this.  When I changed line 172 to "US Navy" I got the error started and that was good, it should error.  However when I changed line 174 to "Lieutenant" it didn't fix the error I got a different error which was:

Warn (yamlValidator):	 Key 'rank_title' at level 'demographics' must be one of the following values: ['Private (Recruit)', 'Private', 'Private First Class', 'Specialist', 'Corporal', 'Sergeant', 'Staff Sergeant', 'Sergeant First Class', 'Master Sergeant', 'First Sergeant', 'Sergeant Major', 'Command Sergeant Major', 'Sergeant Major of the Army', 'Warrant Officer 1', 'Chief Warrant Officer 2', 'Chief Warrant Officer 3', 'Chief Warrant Officer 4', 'Chief Warrant Officer 5', '2nd Lieutenant', '1st Lieutenant', 'Captain', 'Major', 'Lieutenant Colonel', 'Colonel', 'Brigadier General', 'Major General', 'Lieutenant General', 'Army Chief of Staff (special)', 'General', 'Airman Basic', 'Airman', 'Airman First Class', 'Senior Airman', 'Technical Sergeant', 'Senior Master Sergeant', 'First Sergeant / Chief Master Sergeant', 'Chief Master Sergeant of the Air Force', 'Air Force Chief of Staff (special)', 'Seaman Recruit', 'Seaman Apprentice', 'Seaman', 'Petty Officer Third Class', 'Petty Officer Second Class', 'Petty Officer First Class', 'Chief Petty Officer', 'Senior Chief Petty Officer', 'Master Chief Petty Officer', 'Master Chief Petty Officer of the Navy', 'Master Chief Petty Officer of the Coast Guard', 'Chief Warrant Officer', 'Ensign', 'Lieutenant, Junior Grade', 'Lieutenant Commander', 'Commander', 'Rear Admiral (Lower Half)', 'Rear Admiral (Upper Half)', 'Vice Admiral', 'Chief of Naval Operations (special)', 'Commandant of the Coast Guard (special)', 'Admiral', 'Lance Corporal', 'Gunnery Sergeant', 'Master Gunnery Sergeant', 'Sergeant Major of the Marine Corps', 'Warrant Officer', 'Commandant of the Marine Corps', 'Specialist 1', 'Specialist 2', 'Specialist 3', 'Specialist 4', 'Chief Master Sergeant', 'Chief Master Sergeant of the Space Force', 'Chief of Space Operations'] but is 'Lieutenant' instead.
Warn (yamlValidator):	 Key 'unstructured' at level 'Scenes/State' should be type 'string' but is <class 'NoneType'> instead.

So is my enum just not updated?  or did I do something else wrong?

Comment (nextcen-dgemoets) [reply api_files/dependencies.json:73]: I got the same issue, but the problem is with the `api.yaml` file, not the validator or your testing.  It turns out I missed `Lieutenant` in creating all of the valid values for `rank_title`.  It's only used in Coast Guard/Navy (all other Lieutenants have some type of modifier).

Comment (kaitlyn-sharo) [reply api_files/dependencies.json:73]: Is this an action item for me?

**Review by nextcen-dgemoets — COMMENTED** [review]
Are multiple values in any of the fields supposed to be supported?  E.g., can you say, `"military_branch": ["US Army", "US Navy"]` to match either value?  It doesn't appear so, and that's okay.  But it's less clear in the case of the `requirement` field.  Maybe the only action is to clarify in the README where single/multiple values are allowed for the various fields.

Related, when I specified a `military_branch` of `["US Army"]` instead of `"US Army"`, it didn't work.  I had to remove the brackets.  One possibility is to log something (Info? Warn?) if you've specified a list of values if only a single value is supported.

As far as I know, it's not a requirement at the moment to support multiple values for `field1` or `field2`.

File: api_files/dependencies.json (hunk 1 above)
Comment (kaitlyn-sharo) [reply api_files/dependencies.json:73]: Completed

File: validator.py [hunk 2]
```diff
@@ -551,6 +556,54 @@ def simple_value_matching(self):
...
+                    # no index given for an array, skip this key
+                    continue
+                # look for matching keys using possibleParents
+                for req_set in self.dep_json['deepLinks'][parent_key]:
+                    conditions = True
+                    # check if the conditions are true
+                    explanation = ""
+                    for c in req_set['condition']:
+                        conditions = conditions and self.does_key_have_value(p.split('.')+c.split('.'), req_set['condition'][c], copy.deepcopy(self.loaded_yaml))
+                        explanation += "key " + c + " with value " + str(req_set['condition'][c]) + '; '
```
[the remark this answers is not in the document]
Comment (kaitlyn-sharo) [reply validator.py]: Good call. Updated now to show key-value pairs easier too (key: value)

File: README.md [hunk 3]
```diff
@@ -66,4 +66,5 @@ The dependencies json lists specific rules for the validator to follow. When lis
...
 | `simpleAllowedValues` | "If [field1] has a value of [value1], then [field2] values must be [...]" | A dictionary where each key is a [field1] and the value is an object where each key is a possible value for field 1. Those keys are mapped to objects whose keys are [field2] names with a matching value of an array of possible allowed values for field2 |
+| `deepLinks` | "If [field1] has a value of [value1] and [field2] has a value of [value2], then [field3] values must be [...]" | An object where each key is a shared parent of all fields throughout the rest of the object. For example, state.characters[].vitals is the parent of conscious, avpu, and mental_status. Each value contains "sharedParent", "condition" and "requirement".  "condition" is the set of fields and values that must be found in order for "requirement" to be required. "requirement" is a set of fields matched to their allowed values given the conditions |
```
[the remark this answers is not in the document]
Comment (kaitlyn-sharo) [reply README.md]: I changed it to generic letters. Hopefully it still helps

**Comment by kaitlyn-sharo** [comment]
> Are multiple values in any of the fields supposed to be supported? E.g., can you say, `"military_branch": ["US Army", "US Navy"]` to match either value? It doesn't appear so, and that's okay. But it's less clear in the case of the `requirement` field. Maybe the only action is to clarify in the README where single/multiple values are allowed for the various fields.
> 
> Related, when I specified a `military_branch` of `["US Army"]` instead of `"US Army"`, it didn't work. I had to remove the brackets. One possibility is to log something (Info? Warn?) if you've specified a list of values if only a single value is supported.
> 
> As far as I know, it's not a requirement at the moment to support multiple values for `field1` or `field2`.

Hmmm, if I allowed multiple values for field1 and field2 it might be more efficient, and it would certainly make dependencies.json cleaner. I'll give it a try but as of right now, no that's not allowed but requirement can have multiple allowed values. If I can't get the multiple values quickly, since it's not a requirement for now, I'll just update the documentation

**Review by nextcen-dgemoets — COMMENTED** [review]
Looks good except for one missing set of apostrophes at the end:
`Key 'rank_title' must have one of the following values ['Lieutenant', 'Captain'] because 'state.characters[0].demographics' has key-value pairs ('rank': 'O-3'); ('mi
litary_branch': 'US Navy'), but instead value is Corporal`

**Comment by kaitlyn-sharo** [comment]
> Looks good except for one missing set of apostrophes at the end: `Key 'rank_title' must have one of the following values ['Lieutenant', 'Captain'] because 'state.characters[0].demographics' has key-value pairs ('rank': 'O-3'); ('mi litary_branch': 'US Navy'), but instead value is Corporal`

Fixed. Pesky apostrophes...

**Review by nextcen-dgemoets — APPROVED** [review]

**Review by brianpippin — APPROVED** [review]
Everything is working for me now as expected.

**Pull request merged**